From 1f06a75fd5b53ca9281df8486af4ccc18f2725c2 Mon Sep 17 00:00:00 2001
From: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
Date: Mon, 3 Apr 2023 21:52:35 -0700
Subject: [PATCH] ath12k: Fix the transmit completion mask and handle irq
configuration properly

Fix the transmit completion ring4 with the right mask ATH12K_TX_RING_MASK_4,
instead of ATH12K_TX_RING_MASK_3. ATH12K_TX_RING_MASK_3 is used for error
rx release ring.
Flush cache entries before updating the ring indexes

Signed-off-by: Ramanathan Choodamani <quic_rchoodam@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/ahb.c   | 7 +++++--
 drivers/net/wireless/ath/ath12k/core.h  | 1 +
 drivers/net/wireless/ath/ath12k/dp_tx.c | 1 +
 drivers/net/wireless/ath/ath12k/hw.c    | 2 +-
 4 files changed, 8 insertions(+), 3 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/ahb.c b/drivers/net/wireless/ath/ath12k/ahb.c
index c50f2c8..3ed7fb0 100644
--- a/drivers/net/wireless/ath/ath12k/ahb.c
+++ b/drivers/net/wireless/ath/ath12k/ahb.c
@@ -169,7 +169,7 @@ static inline void ath12k_ahb_write32(struct ath12k_base *ab, u32 offset,
 			   ab->mem_ce, offset, value);
 		iowrite32(value, ab->mem_ce + offset);
 	} else {
-		ath12k_dbg(ab, ATH12K_DBG_OFFSET, "mem:%p offset:%x value:%d\n",
+		ath12k_dbg(ab, ATH12K_DBG_OFFSET, "dp mem:%p offset:%x value:%d\n",
 			   ab->mem, offset, value);
 		iowrite32(value, ab->mem + offset);
 	}
@@ -502,7 +502,10 @@ static int ath12k_ahb_config_ext_irq(struct ath12k_base *ab)
 			       ath12k_ahb_ext_grp_napi_poll, NAPI_POLL_WEIGHT);
 
 		for (j = 0; j < ATH12K_EXT_IRQ_NUM_MAX; j++) {
-			if (ab->hw_params->ring_mask->tx[i] & BIT(j)) {
+			if (ab->hw_params->ring_mask->tx[i]  &&
+					(j <= ATH12K_MAX_TCL_RING_NUM) &&
+					(ab->hw_params->ring_mask->tx[i] &
+					 BIT(ab->hw_params->hal_ops->tcl_to_wbm_rbm_map[j].wbm_ring_num))) {
 				irq_grp->irqs[num_irq++] =
 					wbm2host_tx_completions_ring1 - j;
 			}
diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 60b96a0..34a30d9 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -157,6 +157,7 @@ extern bool ath12k_cold_boot_cal;
 
 #define ATH12K_IRQ_NUM_MAX 57
 #define ATH12K_EXT_IRQ_NUM_MAX	16
+#define ATH12K_MAX_TCL_RING_NUM 3
 
 struct ath12k_ext_irq_grp {
 	struct ath12k_base *ab;
diff --git a/drivers/net/wireless/ath/ath12k/dp_tx.c b/drivers/net/wireless/ath/ath12k/dp_tx.c
index 12e4f7d..47f284a 100644
--- a/drivers/net/wireless/ath/ath12k/dp_tx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_tx.c
@@ -491,6 +491,7 @@ int ath12k_dp_tx(struct ath12k *ar, struct ath12k_link_vif *arvif,
 	tcl_cmd->info4 = arvif->desc.info4;
 	tcl_cmd->info5 = 0;
 
+	dsb(st);
 	ath12k_hal_srng_access_umac_src_ring_end_nolock(tcl_ring);
 
 	ath12k_dbg_dump(ab, ATH12K_DBG_DP_TX, NULL, "dp tx msdu: ",
diff --git a/drivers/net/wireless/ath/ath12k/hw.c b/drivers/net/wireless/ath/ath12k/hw.c
index 8809454..012d43b 100644
--- a/drivers/net/wireless/ath/ath12k/hw.c
+++ b/drivers/net/wireless/ath/ath12k/hw.c
@@ -841,7 +841,7 @@ static const struct ath12k_hw_ring_mask ath12k_hw_ring_mask_ipq5332 = {
 		ATH12K_TX_RING_MASK_0,
 		ATH12K_TX_RING_MASK_1,
 		ATH12K_TX_RING_MASK_2,
-		ATH12K_TX_RING_MASK_3,
+		ATH12K_TX_RING_MASK_4,
 		0, 0, 0, 0, 0, 0, 0, 0,
 	},
 	.rx_mon_dest = {
-- 
2.7.4

